src/pubsub.c

```diff
@@ -615,6 +615,107 @@ void publishCommand(client *c) {
     addReplyLongLong(c,receivers);
 }
 
+typedef struct {
+    dict *subscription_dict;
+    long count;
+    sds pat;
+} pubsubSubscribersCmdArgs;
+
+static void pubsubReplySubscribersClientInfo(client *c, list *clients) {
+    listNode *ln;
+    listIter li;
+    listRewind(clients, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        client *client = listNodeValue(ln);
+        addReplyMapLen(c, 3);
+        addReplyBulkCString(c, "name");
+        if (client->name) {
+            addReplyBulk(c, client->name);
+        } else {
+            addReplyBulkCBuffer(c, "", 0);
+        }
+        addReplyBulkCString(c, "id");
+        addReplyBulkLongLong(c, client->id);
+        addReplyBulkCString(c, "addr");
+        addReplyBulkCString(c, getClientPeerId(client));
+    }
+}
+
+static void pubsubReplySubscribers(client *c, pubsubSubscribersCmdArgs args) {
+    void *dl = addReplyDeferredLen(c);
+    long long channel_count = 0;
+    long long subscription_count = 0;
+
+    dictIterator *di = dictGetIterator(args.subscription_dict);
+    dictEntry *de;
+    while((de = dictNext(di)) != NULL && subscription_count < args.count) {
+        /* Filter the channels/subscriptions. */
+        sds channel = ((robj *)dictGetKey(de))->ptr;
+        if (args.pat && !stringmatchlen(args.pat, sdslen(args.pat), channel, sdslen(channel), 0)) {
+            continue;
+        }
+        list *clients = dictGetVal(de);
+        serverAssert(clients != NULL);
+        addReplyBulkCString(c, channel);
+        addReplyArrayLen(c, listLength(clients));
+        pubsubReplySubscribersClientInfo(c, clients);
+        subscription_count += listLength(clients);
```

Comment:
isn't it odd that we stop the loop past the COUNT limit?
if we have one channel with a ton of clients (of a very low limit), we'll pass the limit by a lot.

Reply:
The alternative is we don't give complete client results, which seems worse from a debugging API. I wouldn't expect a single client to be listening to millions of connections, but maybe there are use cases I am not aware of?